c10/core/TensorImpl.h

```diff
@@ -1092,21 +1134,27 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
     refresh_contiguous();
   }
 
+  // Inference tensor doesn't have version counter,
+  // set_version_counter is no-op for them.
```

Comment:
I think you can tighten the invariant here. According to the RFC, inference tensors produce inference views, and normal tensors produce normal views. Thus, an inference tensor which is getting its version counter "set" (we only set version counters when setting up view information) should only ever have gotten the version counter from another inference tensor, which is guaranteed to be disabled. So you should be able to assert that if you ARE an inference tensor, it must be the case that the passed version counter is disabled. And if the passed version counter is disabled, it actually doesn't matter if this is an inference tensor or not, it's always valid to assign the version counter (because overwriting a nullptr with a nullptr does the same thing)--so long as the invariant is upheld.